File: 1986.minimum-number-of-work-sessions-to-finish-the-tasks.py
#
# @lc app=leetcode id=1986 lang=python3
#
# [1986] Minimum Number of Work Sessions to Finish the Tasks
#

# @lc code=start
from typing import List


class Solution:
    def minSessions(self, tasks: List[int], sessionTime: int) -> int:
        n = len(tasks)
        dp = [[n, 0] for i in range(1 << n)]
        dp[-1] = [1, sessionTime]
        for mask in range(1 << n - 1, -1, -1):
            for i in range(n):
                if mask & (1 << i) == 0:
                    cntSession, remainTime = dp[mask ^ (1 << i)]
                    if tasks[i] <= remainTime:
                        remainTime -= tasks[i]
                    else:
                        cntSession += 1
                        remainTime = sessionTime - tasks[i]
                    if cntSession < dp[mask][0] or (cntSession == dp[mask][0] and remainTime > dp[mask][1]):
                        dp[mask] = [cntSession, remainTime]
        return dp[0][0]
# @lc code=end

File: 1986.minimum-number-of-work-sessions-to-finish-the-tasks.cs
/*
 * @lc app=leetcode id=1986 lang=csharp
 *
 * [1986] Minimum Number of Work Sessions to Finish the Tasks
 */

// @lc code=start
using System;

public class Solution {
    public int MinSessions(int[] tasks, int sessionTime) {
        int n = tasks.Length;
        int[,] dp = new int[1 << n, 2];
        for (int i = 0; i < (1 << n); i++) {
            dp[i, 0] = n;
            dp[i, 1] = 0;
        }
        dp[(1 << n) - 1, 0] = 1;
        dp[(1 << n) - 1, 1] = sessionTime;

        for (int mask = (1 << n) - 1; mask >= 0; mask--) {
            for (int i = 0; i < n; i++) {
                if ((mask & (1 << i)) == 0) {
                    int cntSession = dp[mask ^ (1 << i), 0];
                    int remainTime = dp[mask ^ (1 << i), 1];

                    if (tasks[i] <= remainTime) {
                        remainTime -= tasks[i];
                    } else {
                        cntSession += 1;
                        remainTime = sessionTime - tasks[i];
                    }

                    if (cntSession < dp[mask, 0] || (cntSession == dp[mask, 0] && remainTime > dp[mask, 1])) {
                        dp[mask, 0] = cntSession;
                        dp[mask, 1] = remainTime;
                    }
                }
            }
        }

        return dp[0, 0];
    }
}



// @lc code=end

